File: backend/src/bond_monitor/domain/trading/deploy_session.py
"""Ephemeral deploy session — frozen buy/reinvest plan for atomic execution."""

from __future__ import annotations

import secrets
from collections.abc import Sequence
from dataclasses import dataclass, field, replace
from datetime import UTC, date, datetime, timedelta
from typing import Literal

from bond_monitor.domain.bonds.models import BondRecord
from bond_monitor.domain.portfolio.models import Portfolio, PortfolioPosition
from bond_monitor.domain.portfolio.policies import (
    DEFAULT_BOND_SELECTION_POLICY,
    DEFAULT_DURATION_POLICY,
    DEFAULT_PLANNING_POLICY,
    BondSelectionPolicy,
    DurationPolicy,
    PlanningPolicy,
)
from bond_monitor.domain.portfolio.selection import has_usable_price
from bond_monitor.domain.trading.holdings import HoldingView
from bond_monitor.domain.trading.ids import stable_id
from bond_monitor.domain.trading.policies import (
    DeploySessionPolicy,
    buy_limit_price_buffer,
    suggested_buy_limit_price_pct,
)
from bond_monitor.domain.trading.ports import BrokerActiveOrder
from bond_monitor.domain.trading.suggestions import Suggestion

def _as_utc(dt: datetime) -> datetime:
    if dt.tzinfo is None:
        return dt.replace(tzinfo=UTC)
    return dt.astimezone(UTC)


DeploySessionStatus = Literal["active", "completed", "cancelled", "expired"]
DeploySessionItemKind = Literal["buy", "reinvest"]

_TERMINAL_ORDER_STATUSES = frozenset(
    {
        "EXECUTION_REPORT_STATUS_FILL",
        "EXECUTION_REPORT_STATUS_CANCELLED",
        "EXECUTION_REPORT_STATUS_REJECTED",
    }
)
_FILLED_ORDER_STATUS = "EXECUTION_REPORT_STATUS_FILL"


@dataclass
class DeploySessionItem:
    id: str
    kind: DeploySessionItemKind
    isin: str
    name: str
    lots: int
    figi: str | None
    suggested_price_pct: float
    estimated_amount_rub: float
    reason: str
    status: DeploySessionItemStatus = "pending"
    source_isin: str | None = None
    due_date: date | None = None
    order_id: str | None = None
    urgency: Literal["normal", "soon", "critical"] = "normal"

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "kind": self.kind,
            "isin": self.isin,
            "name": self.name,
            "lots": self.lots,
            "figi": self.figi,
            "suggested_price_pct": self.suggested_price_pct,
            "estimated_amount_rub": self.estimated_amount_rub,
            "reason": self.reason,
            "status": self.status,
            "source_isin": self.source_isin,
            "due_date": self.due_date.isoformat() if self.due_date else None,
            "order_id": self.order_id,
            "urgency": self.urgency,
        }

    @classmethod
    def from_dict(cls, data: dict) -> DeploySessionItem:
        due_raw = data.get("due_date")
        return cls(
            id=str(data["id"]),
            kind=data["kind"],  # type: ignore[arg-type]
            isin=str(data["isin"]),
            name=str(data["name"]),
            lots=int(data["lots"]),
            figi=data.get("figi"),
            suggested_price_pct=float(data["suggested_price_pct"]),
            estimated_amount_rub=float(data["estimated_amount_rub"]),
            reason=str(data["reason"]),
            status=data.get("status", "pending"),  # type: ignore[arg-type]
            source_isin=data.get("source_isin"),
            due_date=date.fromisoformat(due_raw) if due_raw else None,
            order_id=data.get("order_id"),
            urgency=data.get("urgency", "normal"),  # type: ignore[arg-type]
        )


@dataclass
class DeploySession:
    id: str
    portfolio_id: str
    status: DeploySessionStatus
    items: list[DeploySessionItem]
    cash_snapshot_rub: float
    created_at: datetime
    expires_at: datetime
    warnings: list[str] = field(default_factory=list)
    completed_at: datetime | None = None

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "portfolio_id": self.portfolio_id,
            "status": self.status,
            "items": [item.to_dict() for item in self.items],
            "cash_snapshot_rub": self.cash_snapshot_rub,
            "created_at": self.created_at.isoformat(),
            "expires_at": self.expires_at.isoformat(),
            "warnings": list(self.warnings),
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
        }

    @classmethod
    def from_dict(cls, data: dict) -> DeploySession:
        completed_raw = data.get("completed_at")
        return cls(
            id=str(data["id"]),
            portfolio_id=str(data["portfolio_id"]),
            status=data["status"],  # type: ignore[arg-type]
            items=[DeploySessionItem.from_dict(item) for item in data.get("items", [])],
            cash_snapshot_rub=float(data["cash_snapshot_rub"]),
            created_at=_as_utc(datetime.fromisoformat(str(data["created_at"]))),
            expires_at=_as_utc(datetime.fromisoformat(str(data["expires_at"]))),
            warnings=list(data.get("warnings") or []),
            completed_at=(
                _as_utc(datetime.fromisoformat(completed_raw)) if completed_raw else None
            ),
        )


@dataclass(frozen=True)
class DeploySessionProgress:
    total: int
    pending: int
    placed: int
    filled: int
    skipped: int
    stale: int


def deploy_session_progress(session: DeploySession) -> DeploySessionProgress:
    counts = {status: 0 for status in ("pending", "placed", "filled", "skipped", "stale")}
    for item in session.items:
        counts[item.status] = counts.get(item.status, 0) + 1
    return DeploySessionProgress(
        total=len(session.items),
        pending=counts["pending"],
        placed=counts["placed"],
        filled=counts["filled"],
        skipped=counts["skipped"],
        stale=counts["stale"],
    )


def _item_key(suggestion: Suggestion) -> str:
    if suggestion.kind == "reinvest" and suggestion.source_isin and suggestion.due_date:
        return f"{suggestion.source_isin}:{suggestion.due_date.isoformat()}"
    return suggestion.isin


def _estimated_amount_rub(
    suggestion: Suggestion,
    universe_by_isin: dict[str, BondRecord],
) -> float:
    bond = universe_by_isin.get(suggestion.isin)
    if bond is not None and bond.price_per_lot_rub:
        return round(suggestion.lots * bond.price_per_lot_rub, 2)
    if bond is not None:
        face = bond.face_value or 1000.0
        aci = bond.accrued_interest or 0.0
        price_pct = suggestion.suggested_price_pct or 100.0
        lot_size = bond.lot_size or 1
        return round(
            suggestion.lots * lot_size * (face * price_pct / 100.0 + aci),
            2,
        )
    return 0.0


def suggestion_to_session_item(
    suggestion: Suggestion,
    *,
    session_id: str,
    portfolio_id: str,
    universe_by_isin: dict[str, BondRecord],
) -> DeploySessionItem:
    item_id = stable_id(
        portfolio_id,
        "deploy-item",
        f"{session_id}:{suggestion.kind}:{_item_key(suggestion)}",
    )
    return DeploySessionItem(
        id=item_id,
        kind=suggestion.kind,  # type: ignore[arg-type]
        isin=suggestion.isin,
        name=suggestion.name,
        lots=suggestion.lots,
        figi=suggestion.figi,
        suggested_price_pct=float(suggestion.suggested_price_pct or 100.0),
        estimated_amount_rub=_estimated_amount_rub(suggestion, universe_by_isin),
        reason=suggestion.reason,
        source_isin=suggestion.source_isin,
        due_date=suggestion.due_date,
        urgency=suggestion.urgency,
    )


def build_deploy_session_plan(
    portfolio: Portfolio,
    holdings: list[HoldingView],
    positions: list[PortfolioPosition],
    universe: Sequence[BondRecord],
    *,
    available_cash: float,
    today: date,
    key_rate: float,
    tax_rate: float,
    selection_policy: BondSelectionPolicy = DEFAULT_BOND_SELECTION_POLICY,
    planning_policy: PlanningPolicy = DEFAULT_PLANNING_POLICY,
    duration_policy: DurationPolicy = DEFAULT_DURATION_POLICY,
    policy: DeploySessionPolicy = DeploySessionPolicy(),
    now: datetime | None = None,
    session_id: str | None = None,
) -> DeploySession:
    """Собрать снимок buy+reinvest рекомендаций для фиксации в сессии."""
    from bond_monitor.domain.trading.advisory import (
        build_buy_suggestions,
        build_reinvest_suggestions,
    )

    universe_by_isin = {bond.isin: bond for bond in universe}
    resolved_session_id = session_id or secrets.token_hex(16)
    created = now or datetime.now(UTC)
    expires = created + timedelta(hours=policy.ttl_hours)

    buy_suggestions = build_buy_suggestions(
        portfolio,
        holdings,
        universe,
        universe_by_isin,
        available_cash=available_cash,
        today=today,
        key_rate=key_rate,
        tax_rate=tax_rate,
        duration_policy=duration_policy,
    )
    reinvest_suggestions = build_reinvest_suggestions(
        portfolio,
        positions,
        universe,
        today=today,
        key_rate=key_rate,
        tax_rate=tax_rate,
        policy=selection_policy,
        planning=planning_policy,
        duration_policy=duration_policy,
    )
    items = [
        suggestion_to_session_item(
            suggestion,
            session_id=resolved_session_id,
            portfolio_id=portfolio.id,
            universe_by_isin=universe_by_isin,
        )
        for suggestion in buy_suggestions + reinvest_suggestions
    ]
    return DeploySession(
        id=resolved_session_id,
        portfolio_id=portfolio.id,
        status="active",
        items=items,
        cash_snapshot_rub=available_cash,
        created_at=_as_utc(created),
        expires_at=_as_utc(expires),
    )


def session_item_to_suggestion(
    item: DeploySessionItem,
    universe_by_isin: dict[str, BondRecord],
) -> Suggestion:
    bond = universe_by_isin.get(item.isin)
    market_price = bond.last_price if bond is not None and bond.last_price else None
    return Suggestion(
        id=item.id,
        kind=item.kind,
        isin=item.isin,
        name=item.name,
        lots=item.lots,
        figi=item.figi,
        suggested_price_pct=item.suggested_price_pct,
        market_price_pct=market_price,
        reason=item.reason,
        due_date=item.due_date,
        source_isin=item.source_isin,
        urgency=item.urgency,
    )


def session_items_to_suggestions(
    session: DeploySession,
    universe: Sequence[BondRecord],
    *,
    kinds: set[DeploySessionItemKind],
) -> list[Suggestion]:
    """Преобразовать pending-позиции сессии в actionable suggestions."""
    universe_by_isin = {bond.isin: bond for bond in universe}
    result: list[Suggestion] = []
    for item in session.items:
        if item.kind not in kinds:
            continue
        if item.status != "pending":
            continue
        result.append(session_item_to_suggestion(item, universe_by_isin))
    return result


def _implied_market_price_pct(
    suggested_price_pct: float,
    account_kind,
) -> float:
    buffer = buy_limit_price_buffer(account_kind)
    return suggested_price_pct / (1 + buffer)


def apply_session_staleness(
    session: DeploySession,
    universe: Sequence[BondRecord],
    *,
    portfolio: Portfolio,
    policy: DeploySessionPolicy = DeploySessionPolicy(),
    now: datetime | None = None,
) -> DeploySession:
    """Проверить актуальность цен и доступность бумаг."""
    now = _as_utc(now or datetime.now(UTC))
    if _as_utc(session.expires_at) <= now:
        return replace(session, status="expired")

    universe_by_isin = {bond.isin: bond for bond in universe}
    warnings: list[str] = list(session.warnings)
    updated_items: list[DeploySessionItem] = []

    for item in session.items:
        if item.status in ("filled", "skipped", "stale", "placed"):
            updated_items.append(item)
            continue

        if item.kind == "reinvest" and item.due_date is not None:
            if item.due_date > now.date():
                updated_items.append(replace(item, status="stale"))
                warnings.append(
                    f"{item.name}: реинвестиция доступна с "
                    f"{item.due_date.strftime('%d.%m.%Y')} — обновите план ближе к дате"
                )
                continue
            if item.due_date < now.date():
                updated_items.append(replace(item, status="stale"))
                warnings.append(
                    f"{item.name}: погашение источника {item.due_date.strftime('%d.%m.%Y')} "
                    "прошло — обновите план"
                )
                continue

        bond = universe_by_isin.get(item.isin)
        if bond is None or not has_usable_price(bond):
            updated_items.append(replace(item, status="stale"))
            warnings.append(f"{item.isin}: бумага недоступна для покупки")
            continue

        implied_market = _implied_market_price_pct(
            item.suggested_price_pct,
            portfolio.account_kind,
        )
        current_market = bond.last_price or implied_market
        if implied_market <= 0:
            updated_items.append(item)
            continue

        drift_pct = abs(current_market - implied_market) / implied_market * 100.0
        if drift_pct >= policy.price_drift_stale_pct:
            updated_items.append(replace(item, status="stale"))
            warnings.append(
                f"{item.name}: цена ушла на {drift_pct:.1f}% — обновите план"
            )
        else:
            if drift_pct >= policy.price_drift_warn_pct:
                warnings.append(
                    f"{item.name}: цена изменилась на {drift_pct:.1f}%"
                )
            updated_items.append(item)

    deduped_warnings = list(dict.fromkeys(warnings))
    return replace(session, items=updated_items, warnings=deduped_warnings)


def sync_session_with_orders(
    session: DeploySession,
    active_orders: Sequence[BrokerActiveOrder],
) -> DeploySession:
    """Синхронизировать статусы items с активными заявками брокера."""
    orders_by_id = {order.order_id: order for order in active_orders}
    updated_items: list[DeploySessionItem] = []

    for item in session.items:
        if item.status != "placed" or not item.order_id:
            updated_items.append(item)
            continue

        order = orders_by_id.get(item.order_id)
        if order is None:
            updated_items.append(replace(item, status="filled"))
            continue

        if order.status == _FILLED_ORDER_STATUS:
            updated_items.append(replace(item, status="filled"))
        elif order.status in _TERMINAL_ORDER_STATUSES:
            updated_items.append(replace(item, status="pending", order_id=None))
        else:
            updated_items.append(item)

    session = replace(session, items=updated_items)
    return complete_session_if_no_pending(session)


def mark_item_placed(
    session: DeploySession,
    item_id: str,
    order_id: str,
) -> DeploySession:
    updated_items: list[DeploySessionItem] = []
    for item in session.items:
        if item.id == item_id:
            updated_items.append(
                replace(item, status="placed", order_id=order_id)
            )
        else:
            updated_items.append(item)
    session = replace(session, items=updated_items)
    return complete_session_if_no_pending(session)


def mark_item_skipped(session: DeploySession, item_id: str) -> DeploySession:
    updated_items: list[DeploySessionItem] = []
    for item in session.items:
        if item.id == item_id:
            updated_items.append(replace(item, status="skipped"))
        else:
            updated_items.append(item)
    session = replace(session, items=updated_items)
    return complete_session_if_no_pending(session)


def find_session_item(session: DeploySession, item_id: str) -> DeploySessionItem | None:
    for item in session.items:
        if item.id == item_id:
            return item
    return None


def complete_session_if_no_pending(session: DeploySession) -> DeploySession:
    """Завершить сессию, когда не осталось pending-позиций (placed = сделано для плана)."""
    if session.status != "active" or not session.items:
        return session
    if any(item.status == "pending" for item in session.items):
        return session
    return replace(
        session,
        status="completed",
        completed_at=datetime.now(UTC),
    )


def session_has_pending_items(session: DeploySession) -> bool:
    return any(item.status == "pending" for item in session.items)


def is_session_active(session: DeploySession, *, now: datetime | None = None) -> bool:
    now = _as_utc(now or datetime.now(UTC))
    return session.status == "active" and _as_utc(session.expires_at) > now

File: backend/internal/domain/trading/deploy_session.go
package trading

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"time"

	"github.com/tonatos/instrumenta/backend/internal/domain/bonds"
	"github.com/tonatos/instrumenta/backend/internal/domain/portfolio"
	"github.com/tonatos/instrumenta/backend/internal/domain/shared"
)

type DeploySessionStatus string

const (
	DeploySessionActive    DeploySessionStatus = "active"
	DeploySessionCompleted DeploySessionStatus = "completed"
	DeploySessionCancelled DeploySessionStatus = "cancelled"
	DeploySessionExpired   DeploySessionStatus = "expired"
)

type DeploySessionItemKind string

const (
	DeploySessionItemBuy      DeploySessionItemKind = "buy"
	DeploySessionItemReinvest DeploySessionItemKind = "reinvest"
)

type DeploySessionItemStatus string

const (
	ItemStatusPending DeploySessionItemStatus = "pending"
	ItemStatusPlaced  DeploySessionItemStatus = "placed"
	ItemStatusFilled  DeploySessionItemStatus = "filled"
	ItemStatusSkipped DeploySessionItemStatus = "skipped"
	ItemStatusStale   DeploySessionItemStatus = "stale"
)

var terminalOrderStatuses = map[string]bool{
	"EXECUTION_REPORT_STATUS_FILL":      true,
	"EXECUTION_REPORT_STATUS_CANCELLED": true,
	"EXECUTION_REPORT_STATUS_REJECTED":  true,
}

const filledOrderStatus = "EXECUTION_REPORT_STATUS_FILL"

type DeploySessionItem struct {
	ID                  string
	Kind                DeploySessionItemKind
	ISIN                string
	Name                string
	Lots                int
	FIGI                *string
	SuggestedPricePct   float64
	EstimatedAmountRub  float64
	Reason              string
	Status              DeploySessionItemStatus
	SourceISIN          *string
	DueDate             *time.Time
	OrderID             *string
	Urgency             SuggestionUrgency
}

type DeploySession struct {
	ID              string
	PortfolioID     string
	Status          DeploySessionStatus
	Items           []DeploySessionItem
	CashSnapshotRub float64
	CreatedAt       time.Time
	ExpiresAt       time.Time
	Warnings        []string
	CompletedAt     *time.Time
}

type DeploySessionProgress struct {
	Total   int
	Pending int
	Placed  int
	Filled  int
	Skipped int
	Stale   int
}

func asUTC(t time.Time) time.Time {
	if t.Location() == time.UTC {
		return t
	}
	return t.UTC()
}

func DeploySessionProgressOf(session DeploySession) DeploySessionProgress {
	counts := map[DeploySessionItemStatus]int{
		ItemStatusPending: 0, ItemStatusPlaced: 0, ItemStatusFilled: 0,
		ItemStatusSkipped: 0, ItemStatusStale: 0,
	}
	for _, item := range session.Items {
		counts[item.Status]++
	}
	return DeploySessionProgress{
		Total:   len(session.Items),
		Pending: counts[ItemStatusPending],
		Placed:  counts[ItemStatusPlaced],
		Filled:  counts[ItemStatusFilled],
		Skipped: counts[ItemStatusSkipped],
		Stale:   counts[ItemStatusStale],
	}
}

func itemKey(s Suggestion) string {
	if s.Kind == SuggestionKindReinvest && s.SourceISIN != nil && s.DueDate != nil {
		return *s.SourceISIN + ":" + shared.FormatISODate(*s.DueDate)
	}
	return s.ISIN
}

func estimatedAmountRub(s Suggestion, universeByISIN map[string]bonds.BondRecord) float64 {
	bond, ok := universeByISIN[s.ISIN]
	if !ok {
		return 0
	}
	if p := bond.PricePerLotRub(); p != nil && *p > 0 {
		return round2(float64(s.Lots) * *p)
	}
	face := bond.FaceValue
	if face <= 0 {
		face = 1000
	}
	aci := 0.0
	if bond.AccruedInterest != nil {
		aci = *bond.AccruedInterest
	}
	pricePct := 100.0
	if s.SuggestedPricePct != nil {
		pricePct = *s.SuggestedPricePct
	}
	lotSize := bond.LotSize
	if lotSize <= 0 {
		lotSize = 1
	}
	return round2(float64(s.Lots*lotSize) * (face*pricePct/100 + aci))
}

func round2(v float64) float64 {
	return float64(int(v*100+0.5)) / 100
}

func SuggestionToSessionItem(
	s Suggestion,
	sessionID, portfolioID string,
	universeByISIN map[string]bonds.BondRecord,
) DeploySessionItem {
	itemID := StableID(portfolioID, "deploy-item", sessionID+":"+string(s.Kind)+":"+itemKey(s))
	price := 100.0
	if s.SuggestedPricePct != nil {
		price = *s.SuggestedPricePct
	}
	kind := DeploySessionItemBuy
	if s.Kind == SuggestionKindReinvest {
		kind = DeploySessionItemReinvest
	}
	return DeploySessionItem{
		ID:                 itemID,
		Kind:               kind,
		ISIN:               s.ISIN,
		Name:               s.Name,
		Lots:               s.Lots,
		FIGI:               s.FIGI,
		SuggestedPricePct:  price,
		EstimatedAmountRub: estimatedAmountRub(s, universeByISIN),
		Reason:             s.Reason,
		SourceISIN:         s.SourceISIN,
		DueDate:            s.DueDate,
		Urgency:            s.Urgency,
		Status:             ItemStatusPending,
	}
}

// BuildDeploySessionPlan freezes buy+reinvest suggestions into a deploy session.
func BuildDeploySessionPlan(
	p portfolio.Portfolio,
	holdings []HoldingView,
	positions []portfolio.PortfolioPosition,
	universe []bonds.BondRecord,
	availableCash float64,
	today time.Time,
	keyRate, taxRate float64,
	selectionPolicy portfolio.BondSelectionPolicy,
	planningPolicy portfolio.PlanningPolicy,
	durationPolicy portfolio.DurationPolicy,
	policy DeploySessionPolicy,
	now *time.Time,
	sessionID *string,
) DeploySession {
	universeByISIN := make(map[string]bonds.BondRecord, len(universe))
	for _, bond := range universe {
		universeByISIN[bond.ISIN] = bond
	}
	resolvedSessionID := randomSessionID()
	if sessionID != nil {
		resolvedSessionID = *sessionID
	}
	created := time.Now().UTC()
	if now != nil {
		created = asUTC(*now)
	}
	expires := created.Add(time.Duration(policy.TTLHours) * time.Hour)

	buySuggestions := BuildBuySuggestions(p, holdings, universe, availableCash, today, keyRate, taxRate, durationPolicy)
	reinvestSuggestions := BuildReinvestSuggestions(
		p, positions, universe, today, keyRate, taxRate, selectionPolicy, planningPolicy, durationPolicy, nil,
	)
	var items []DeploySessionItem
	for _, s := range append(buySuggestions, reinvestSuggestions...) {
		items = append(items, SuggestionToSessionItem(s, resolvedSessionID, p.ID, universeByISIN))
	}
	return DeploySession{
		ID:              resolvedSessionID,
		PortfolioID:     p.ID,
		Status:          DeploySessionActive,
		Items:           items,
		CashSnapshotRub: availableCash,
		CreatedAt:       created,
		ExpiresAt:       expires,
	}
}

func randomSessionID() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	return hex.EncodeToString(b[:])
}

func SessionItemToSuggestion(item DeploySessionItem, universeByISIN map[string]bonds.BondRecord) Suggestion {
	var marketPrice *float64
	if bond, ok := universeByISIN[item.ISIN]; ok && bond.LastPrice != nil {
		marketPrice = bond.LastPrice
	}
	kind := SuggestionKindBuy
	if item.Kind == DeploySessionItemReinvest {
		kind = SuggestionKindReinvest
	}
	price := item.SuggestedPricePct
	return Suggestion{
		ID:                item.ID,
		Kind:              kind,
		ISIN:              item.ISIN,
		Name:              item.Name,
		Lots:              item.Lots,
		FIGI:              item.FIGI,
		SuggestedPricePct: &price,
		MarketPricePct:    marketPrice,
		Reason:            item.Reason,
		DueDate:           item.DueDate,
		SourceISIN:        item.SourceISIN,
		Urgency:           item.Urgency,
	}
}

// SessionItemsToSuggestions converts pending session items to actionable suggestions.
func SessionItemsToSuggestions(session DeploySession, universe []bonds.BondRecord, kinds map[DeploySessionItemKind]bool) []Suggestion {
	universeByISIN := make(map[string]bonds.BondRecord, len(universe))
	for _, bond := range universe {
		universeByISIN[bond.ISIN] = bond
	}
	var result []Suggestion
	for _, item := range session.Items {
		if !kinds[item.Kind] || item.Status != ItemStatusPending {
			continue
		}
		result = append(result, SessionItemToSuggestion(item, universeByISIN))
	}
	return result
}

func impliedMarketPricePct(suggestedPricePct float64, accountKind *AccountKind) float64 {
	buffer := BuyLimitPriceBuffer(accountKind)
	return suggestedPricePct / (1 + buffer)
}

// ApplySessionStaleness marks items stale when prices drift or reinvest timing is wrong.
func ApplySessionStaleness(
	session DeploySession,
	universe []bonds.BondRecord,
	p portfolio.Portfolio,
	policy DeploySessionPolicy,
	now *time.Time,
) DeploySession {
	current := time.Now().UTC()
	if now != nil {
		current = asUTC(*now)
	}
	if !asUTC(session.ExpiresAt).After(current) {
		session.Status = DeploySessionExpired
		return session
	}
	universeByISIN := make(map[string]bonds.BondRecord, len(universe))
	for _, bond := range universe {
		universeByISIN[bond.ISIN] = bond
	}
	warnings := append([]string(nil), session.Warnings...)
	var updated []DeploySessionItem
	for _, item := range session.Items {
		if item.Status == ItemStatusFilled || item.Status == ItemStatusSkipped ||
			item.Status == ItemStatusStale || item.Status == ItemStatusPlaced {
			updated = append(updated, item)
			continue
		}
		if item.Kind == DeploySessionItemReinvest && item.DueDate != nil {
			due := shared.DateOnly(*item.DueDate)
			today := shared.DateOnly(current)
			if due.After(today) {
				item.Status = ItemStatusStale
				warnings = append(warnings, fmt.Sprintf(
					"%s: реинвестиция доступна с %s — обновите план ближе к дате",
					item.Name, shared.FormatDate(&due),
				))
				updated = append(updated, item)
				continue
			}
			if due.Before(today) {
				item.Status = ItemStatusStale
				warnings = append(warnings, fmt.Sprintf(
					"%s: погашение источника %s прошло — обновите план",
					item.Name, shared.FormatDate(&due),
				))
				updated = append(updated, item)
				continue
			}
		}
		bond, ok := universeByISIN[item.ISIN]
		if !ok || !portfolio.HasUsablePrice(bond) {
			item.Status = ItemStatusStale
			warnings = append(warnings, item.ISIN+": бумага недоступна для покупки")
			updated = append(updated, item)
			continue
		}
		implied := impliedMarketPricePct(item.SuggestedPricePct, p.AccountKind)
		currentMarket := implied
		if bond.LastPrice != nil {
			currentMarket = *bond.LastPrice
		}
		if implied <= 0 {
			updated = append(updated, item)
			continue
		}
		driftPct := abs(currentMarket-implied) / implied * 100
		if driftPct >= policy.PriceDriftStalePct {
			item.Status = ItemStatusStale
			warnings = append(warnings, fmt.Sprintf("%s: цена ушла на %.1f%% — обновите план", item.Name, driftPct))
		} else if driftPct >= policy.PriceDriftWarnPct {
			warnings = append(warnings, fmt.Sprintf("%s: цена изменилась на %.1f%%", item.Name, driftPct))
		}
		updated = append(updated, item)
	}
	session.Items = updated
	session.Warnings = dedupeWarnings(warnings)
	return session
}

func abs(v float64) float64 {
	if v < 0 {
		return -v
	}
	return v
}

func dedupeWarnings(warnings []string) []string {
	seen := make(map[string]struct{})
	var out []string
	for _, w := range warnings {
		if _, ok := seen[w]; ok {
			continue
		}
		seen[w] = struct{}{}
		out = append(out, w)
	}
	return out
}

// SyncSessionWithOrders updates placed items from broker order statuses.
func SyncSessionWithOrders(session DeploySession, activeOrders []BrokerActiveOrder) DeploySession {
	ordersByID := make(map[string]BrokerActiveOrder, len(activeOrders))
	for _, order := range activeOrders {
		ordersByID[order.OrderID] = order
	}
	var updated []DeploySessionItem
	for _, item := range session.Items {
		if item.Status != ItemStatusPlaced || item.OrderID == nil {
			updated = append(updated, item)
			continue
		}
		order, ok := ordersByID[*item.OrderID]
		if !ok {
			item.Status = ItemStatusFilled
			updated = append(updated, item)
			continue
		}
		switch {
		case order.Status == filledOrderStatus:
			item.Status = ItemStatusFilled
		case terminalOrderStatuses[order.Status]:
			item.Status = ItemStatusPending
			item.OrderID = nil
		}
		updated = append(updated, item)
	}
	session.Items = updated
	return CompleteSessionIfNoPending(session)
}

func MarkItemPlaced(session DeploySession, itemID, orderID string) DeploySession {
	for i := range session.Items {
		if session.Items[i].ID == itemID {
			session.Items[i].Status = ItemStatusPlaced
			session.Items[i].OrderID = &orderID
			break
		}
	}
	return CompleteSessionIfNoPending(session)
}

func MarkItemSkipped(session DeploySession, itemID string) DeploySession {
	for i := range session.Items {
		if session.Items[i].ID == itemID {
			session.Items[i].Status = ItemStatusSkipped
			break
		}
	}
	return CompleteSessionIfNoPending(session)
}

func FindSessionItem(session DeploySession, itemID string) *DeploySessionItem {
	for i := range session.Items {
		if session.Items[i].ID == itemID {
			return &session.Items[i]
		}
	}
	return nil
}

func CompleteSessionIfNoPending(session DeploySession) DeploySession {
	if session.Status != DeploySessionActive || len(session.Items) == 0 {
		return session
	}
	for _, item := range session.Items {
		if item.Status == ItemStatusPending {
			return session
		}
	}
	session.Status = DeploySessionCompleted
	now := time.Now().UTC()
	session.CompletedAt = &now
	return session
}

func SessionHasPendingItems(session DeploySession) bool {
	for _, item := range session.Items {
		if item.Status == ItemStatusPending {
			return true
		}
	}
	return false
}

func IsSessionActive(session DeploySession, now *time.Time) bool {
	current := time.Now().UTC()
	if now != nil {
		current = asUTC(*now)
	}
	return session.Status == DeploySessionActive && asUTC(session.ExpiresAt).After(current)
}
